util/feed.go

```diff
@@ -0,0 +1,396 @@
+// clean up any entries which have been read by all subscribers.
+//
+// However, in order to improve memory performance each feedEntry maintains a
+// reference counter to determine when it can be safely returned to a shared
+// pool.
+type feedEntry struct {
+	entryType      entryType
+	value          interface{}
+	unsubscribedID int
+	next           chan *feedEntry
+
+	// consumerCount is the number of subscribers still waiting to receive this
+	// entry. This is a reference counter used to determine when entries can be
+	// released to the entry pool; however, an entry is not released when its
+	// own consumerCount is zero, but rather when the consumerCount of the
+	// *next* entry in the linked channel is zero. When the next entry has zero
+	// remaining consumers, that signifies that no more consumers are listening
+	// on the next channel of this entry, and thus it can be released.
+	consumerCount int
+}
+
+type entryType int
+
+const (
+	valueEntry entryType = iota
+	unsubEntry
+	closedEntry
+)
+
+var feedEntryPool = sync.Pool{
+	New: func() interface{} {
+		return &feedEntry{
+			next: make(chan *feedEntry, 1),
+		}
+	},
+}
+
+// A Subscription is used to receive events from a specific Feed. A Subscription
+// should only be instantiated via a call to a Feed's Subscribe() method. Once
+// created, events can be read directly from the Events channel provided by this
+// structure.
+//
+// An example of a typical usage of a Subscription:
+//
+//		subscriber := feed.Subscribe()
+//		for {
+//			select {
+//			case event, ok := <-subscriber.Events:
+//				// Process event...
+//			}
+//		}
+//
+// A Subscription cannot block other Subscriptions to the same feed, and each
+// Subscription will receive all events published by the Feed. A user of a
+// Subscription should not modify Events received over the channel.
+//
+// A Subscription can be closed via the Unsubscribe() method.
+//
+// If the Feed is stopped, the Events channel will be closed after reading all
+// remaining events that were published to the Feed before it was stopped.
+type Subscription struct {
+	Events <-chan interface{}
+	closer chan<- struct{}
+}
+
+// Unsubscribe immediately stops the given Subscriber. After this method is
+// called, the Events channel will be closed.
+func (s *Subscription) Unsubscribe() {
+	close(s.closer)
+}
+
+// A Feed is used to publish a stream of events to a set of Subscribers.  Events
+// are values of arbitrary type, and are received and published as an empty
+// interface.  Each Subscriber will receive, in order, each entry published to
+// the Feed.
+//
+// Entries are published by via the Publish method of the Feed; this cannot be
+// blocked by Subscriber activities and will always return quickly.  Subscribers
+// receive events by reading from their Events channel.
+//
+// The Feed will continue running until its associated Stopper is closed.
+//
+//		func FeedExample() {
```

Comment:
Consider putting this in feed_example_test.go (and structuring it so that godoc and go test recognize it as an example test), so the test suite will ensure that it works as written.